### PR TITLE
Clip the values of the nodes 

### PR DESCRIPTION
**Pull request by samronsin**
Clip the values of the nodes  for constrained trees instead of checking the bounds for all including unconstrained features.

### CODE REVIEW

File: sklearn/tree/_criterion.pyx [hunk 1]
```diff
@@ -245,16 +248,11 @@ cdef class Criterion:
                 (sum_left <= upper_bound * weighted_n_left) &
```
Comment (jjerphan) [inline sklearn/tree/_criterion.pyx:248]: I can't comment on line 237, but can we early return `1` when `monotonic_cst == 0`?

File: sklearn/tree/_criterion.pyx [hunk 2]
```diff
@@ -579,6 +577,18 @@ cdef class ClassificationCriterion(Criterion):
...
+        """Clip the value in dest between lower_bound and upper_bound for monotonic constraints.
+
+        Assumes binary classification and single output, as supported by monotonic constraints.
+        """
+        if dest[0] < lower_bound:
+            dest[0] = lower_bound
+            dest[1] = 1 - lower_bound
+        elif dest[0] > upper_bound:
+            dest[0] = upper_bound
+            dest[1] = 1 - upper_bound
```
Comment (jjerphan) [inline sklearn/tree/_criterion.pyx]:
```suggestion 
        if dest[0] < lower_bound:
            dest[0] = lower_bound
        elif dest[0] > upper_bound:
            dest[0] = upper_bound

        # Class proportions for binary classification must sum to 1.
        dest[1] = 1 - dest[0]
```

File: sklearn/tree/_criterion.pyx [hunk 3]
```diff
@@ -1030,6 +1040,13 @@ cdef class RegressionCriterion(Criterion):
         for k in range(self.n_outputs):
             dest[k] = self.sum_total[k] / self.weighted_n_node_samples
 
+    cdef inline void clip_node_value(self, double* dest, double lower_bound, double upper_bound) noexcept nogil:
+        """Clip the value in dest between lower_bound and upper_bound for monotonic constraints."""
+        if dest[0] < lower_bound:
```
Comment (jjerphan) [inline sklearn/tree/_criterion.pyx:1045]:
```suggestion
        # NOTE: when multi-outputs are supported for monotonic constraints,
        # value of each output must also be clipped.
        if dest[0] < lower_bound:
```

File: sklearn/tree/tests/test_monotonic_tree.py [hunk 4]
```diff
@@ -309,8 +309,8 @@ def assert_nd_reg_tree_children_monotonic_bounded(tree_, monotonic_cst):
     for i in range(tree_.node_count):
         feature = tree_.feature[i]
         node_value = tree_.value[i][0][0]  # unpack value from nx1x1 array
-        assert node_value <= upper_bound[i]
-        assert node_value >= lower_bound[i]
+        assert np.float32(node_value) <= np.float32(upper_bound[i])
+        assert np.float32(node_value) >= np.float32(lower_bound[i])
```
Comment (jjerphan) [inline sklearn/tree/tests/test_monotonic_tree.py:313]:
```suggestion
        # Conversion to float32 is used to handle numerical imprecision.
        assert np.float32(node_value) <= np.float32(upper_bound[i])
        assert np.float32(node_value) >= np.float32(lower_bound[i])
```

**Review by jjerphan — COMMENTED** [review]
Here are a few comments to provide more context. 

Side-note: I think I will look at replacing `value` with a memoryview since manually using a pointer and strides is error-prone and unclear. This is orthogonal to this work and is not required to have this merge, but I think this should make its implementation more readable and maintainable then.

**Comment by samronsin** [comment]
The performance difference of this branch vs `main` at 2131264 is insignificant based a quick test with 
```
n_samples = 100000
n_features = 100
X, y = make_regression(
        n_samples=n_samples,
        n_features=n_features,
        n_informative=n_features // 10,
        noise=50,
        random_state=0,
    )
```
trained with both `DecisionTreeRegressor` and `ExtraTreeRegressor`, using both `BestFirstTreeBuilder` and `DepthFirstTreeBuilder`.

**Pull request merged**